File: PHP/One/Transforms/DetectMethodCalls.php
<?php

namespace One\Transforms\DetectMethodCalls;

use One\AstTransformer\AstTransformer;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\UnresolvedCallExpression;
use One\Ast\Expressions\PropertyAccessExpression;
use One\Ast\Expressions\UnresolvedMethodCallExpression;

class DetectMethodCalls extends AstTransformer {
    function __construct() {
        parent::__construct("DetectMethodCalls");
        
    }
    
    protected function visitExpression($expr) {
        $expr = parent::visitExpression($expr);
        if ($expr instanceof UnresolvedCallExpression && $expr->func instanceof PropertyAccessExpression)
            return new UnresolvedMethodCallExpression($expr->func->object, $expr->func->propertyName, $expr->typeArgs, $expr->args);
        return $expr;
    }
}

File: CSharp/One/Transforms/DetectMethodCalls.cs
using One.Ast;
using One;

namespace One.Transforms
{
    public class DetectMethodCalls : AstTransformer
    {
        public DetectMethodCalls(): base("DetectMethodCalls")
        {
            
        }
        
        protected override Expression visitExpression(Expression expr)
        {
            expr = base.visitExpression(expr);
            if (expr is UnresolvedCallExpression unrCallExpr && unrCallExpr.func is PropertyAccessExpression propAccExpr)
                return new UnresolvedMethodCallExpression(propAccExpr.object_, propAccExpr.propertyName, unrCallExpr.typeArgs, unrCallExpr.args);
            return expr;
        }
    }
}